packages/backend-for-frontend/src/services/toolService.ts

```diff
@@ -0,0 +1,469 @@
+  ctx: WithLogger<BffAppContext>
+): Promise<
+  SuccessfulValidation<ApiKey | ConsumerKey> | FailedValidation<ErrorCodes>
+> {
+  if (!jwt.payload.purposeId) {
+    return {
+      data: undefined,
+      errors: [purposeIdNotFoundInClientAssertion()],
+    };
+  }
+  const keyWithClient = await authorizationClient.token
+    .getKeyWithClientByKeyId({
+      params: {
+        clientId: jwt.payload.sub,
+        keyId: jwt.header.kid,
+      },
+      headers: ctx.headers,
+    })
+    .catch(() => undefined);
+
+  if (!keyWithClient) {
+    return {
+      data: undefined,
+      errors: [
+        clientAssertionPublicKeyNotFound(jwt.header.kid, jwt.payload.sub),
+      ],
+    };
+  }
+
+  assertIsConsumer(ctx.authData.organizationId, keyWithClient);
+
+  const { encodedPem } = await authorizationClient.client.getClientKeyById({
+    headers: ctx.headers,
+    params: {
+      clientId: keyWithClient.client.id,
+      keyId: jwt.header.kid,
+    },
+  });
+
+  const purposeId = unsafeBrandId<PurposeId>(jwt.payload.purposeId);
+
+  if (keyWithClient.client.kind === authorizationApi.ClientKind.enum.API) {
+    return {
+      errors: undefined,
+      data: {
+        clientKind: authorizationApi.ClientKind.enum.API,
+        kid: jwt.header.kid,
+        algorithm: "RS256",
+        publicKey: encodedPem,
+        clientId: unsafeBrandId<ClientId>(jwt.payload.iss),
+        consumerId: unsafeBrandId<TenantId>(keyWithClient.client.consumerId),
+        purposeId,
+      },
+    };
+  }
+
+  const purpose = await purposeProcessClient.getPurpose({
+    params: { id: purposeId },
+    headers: ctx.headers,
+  });
+
+  const agreement = await retrieveAgreement(
+    agreementProcessClient,
+    purpose.consumerId,
+    purpose.eserviceId,
+    ctx
+  );
+
+  const descriptor = await retrieveDescriptor(
+    catalogProcessClient,
+    agreement.eserviceId,
+    agreement.descriptorId,
+    ctx
+  );
+
+  return {
+    errors: undefined,
+    data: {
+      clientKind: authorizationApi.ClientKind.enum.CONSUMER,
+      clientId: unsafeBrandId<ClientId>(jwt.payload.iss),
+      kid: jwt.header.kid,
+      algorithm: "RS256",
+      publicKey: encodedPem,
+      purposeId,
+      consumerId: unsafeBrandId<TenantId>(keyWithClient.client.consumerId),
+      agreementId: unsafeBrandId<AgreementId>(agreement.id),
+      eServiceId: unsafeBrandId<EServiceId>(agreement.eserviceId),
+      agreementState: agreementStateToItemState(agreement.state),
+      purposeState: retrievePurposeItemState(purpose),
+      descriptorState: descriptorStateToItemState(descriptor.state),
+    },
+  };
+}
+
+async function retrieveAgreement(
+  agreementClient: AgreementProcessClient,
+  consumerId: string,
+  eserviceId: string,
+  ctx: WithLogger<BffAppContext>
+): Promise<agreementApi.Agreement> {
+  const agreements = await getAllAgreements(agreementClient, ctx.headers, {
+    consumersIds: [consumerId],
+    eservicesIds: [eserviceId],
+    states: [
```

Comment:
the agreement could also be `ARCHIVED`.
Note that `ACTIVE` and `SUSPENDED` should have the priority, for example if we have
- Agreement 1 `ARCHIVED`
- Agreement 2 `ACTIVE`

the correct state is `ACITVE`

Reply:
I modified the logic to give priority to SUSPENDED or ACTIVE agreements. if i do not find one of either, i simply return the first one.